src/parquet/arrow/writer.cc

```diff
@@ -342,29 +358,297 @@ Status FileWriter::Impl::WriteFlatColumnChunk(
+  // 3 -> list is non-null, element is non-null
+  int16_t valid_element = 3;
+  int16_t null_element = 2;
+  int16_t empty_list = 1;
+  int16_t null_list = 0;
+  if (column_writer->descr()->max_definition_level() != 3) {
+    if (column_writer->descr()->max_definition_level() > 3) {
+      return Status::NotImplemented(
+          "Only primitive arrays with max definition level <= 3 are supported yet");
+    } else if ((column_writer->descr()->max_definition_level() == 2) &&
+               nullable_elements) {
+      // Lists are required
+      valid_element = 2;
+      null_element = 1;
+      empty_list = 0;
+    } else if ((column_writer->descr()->max_definition_level() == 2) && nullable_lists) {
+      // Elements are required
+      valid_element--;
+    } else if ((column_writer->descr()->max_definition_level() == 1) && !nullable_lists &&
+               !nullable_elements) {
+      valid_element = 1;
+      null_element = 1;
+      empty_list = 0;
+      null_list = 0;
+    } else {
+      return Status::NotImplemented(
+          "Primitive arrays with max definition level == 0 aren't supported yet");
+    }
+  }
+
+  // Generate repetition levels
+  Int16Builder rep_levels(pool_, ::arrow::int16());
+  Int16Builder def_levels(pool_, ::arrow::int16());
+  const uint8_t* valid_lists = data->null_bitmap_data();
+  const int32_t* raw_offsets = data->raw_offsets();
+  INIT_BITSET(valid_lists, offset);
+  const uint8_t* valid_values = data->values()->null_bitmap_data();
+  INIT_BITSET(valid_values, raw_offsets[offset]);
+  for (int64_t i = 0; i < length; i++) {
+    rep_levels.Append(0);
+    if (bitset_valid_lists & (1 << bit_offset_valid_lists)) {
+      // not null, so we have offsets
+      int32_t len = raw_offsets[i + 1] - raw_offsets[i];
+      if (len == 0) {
+        def_levels.Append(empty_list);
+      } else {
+        if (bitset_valid_values & (1 << bit_offset_valid_values)) {
+          def_levels.Append(valid_element);
+        } else {
+          def_levels.Append(null_element);
+        }
+        READ_NEXT_BITSET(valid_values);
+        for (int32_t j = 1; j < len; j++) {
+          rep_levels.Append(1);
+          if (bitset_valid_values & (1 << bit_offset_valid_values)) {
+            def_levels.Append(valid_element);
+          } else {
+            def_levels.Append(null_element);
+          }
+          READ_NEXT_BITSET(valid_values);
+        }
+      }
+    } else {
+      def_levels.Append(null_list);
+    }
+    READ_NEXT_BITSET(valid_lists);
+  }
+  std::shared_ptr<Array> def_levels_array;
+  def_levels.Finish(&def_levels_array);
+  const int16_t* def_levels_ptr = reinterpret_cast<const int16_t*>(
+      static_cast<Int16Array*>(def_levels_array.get())->data()->data());
+
+  std::shared_ptr<Array> rep_levels_array;
+  rep_levels.Finish(&rep_levels_array);
+  const int16_t* rep_levels_ptr = reinterpret_cast<const int16_t*>(
+      static_cast<Int16Array*>(rep_levels_array.get())->data()->data());
```

Comment:
Can you factor out the construction of the repetition and definition levels into a separate method? When we are dealing with multiple levels of nested types, this function will eventually need to accept a vector of offsets arrays and a vector of bitmaps